Engine_Revit_UI/Query/CeilingPattern.cs

```diff
@@ -0,0 +1,147 @@
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading.Tasks;
+
+using Autodesk.Revit.DB;
+using BH.Engine.Geometry;
+using BH.oM.Geometry;
+
+namespace BH.UI.Revit.Engine
+{
+    public static partial class Query
+    {
+        public static List<BH.oM.Geometry.Line> CeilingPattern(this Ceiling ceiling, PlanarSurface surface)
+        {
+            CeilingType ceilingType = ceiling.Document.GetElement(ceiling.GetTypeId()) as CeilingType;
+            CompoundStructure comStruct = ceilingType.GetCompoundStructure();
+
+            if (comStruct != null && comStruct.GetLayers().Count > 0)
+                return (ceiling.Document.GetElement(comStruct.GetLayers()[0].MaterialId) as Material).CeilingPattern(surface);
+
+            List<ElementId> materialIds = ceiling.GetMaterialIds(false).ToList();
+            if (materialIds.Count == 0)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Ceiling patterns could not be pulled for ceiling element with the ID " + ceiling.Id);
+                return new List<oM.Geometry.Line>();
+            }
+
+            return (ceiling.Document.GetElement(materialIds[0]) as Material).CeilingPattern(surface);
+        }
+
+        public static List<BH.oM.Geometry.Line> CeilingPattern(this Material revitMaterial, PlanarSurface surface)
+        {
+            BoundingBox box = surface.IBounds();
+            double z = surface.ExternalBoundary.IControlPoints().Max(x => x.Z);
+
+            double yLength = box.Max.Y - box.Min.Y;
+
+            BH.oM.Geometry.Line leftLine = new oM.Geometry.Line
+            {
+                Start = new oM.Geometry.Point { X = box.Min.X, Y = box.Min.Y, Z = z },
+                End = new oM.Geometry.Point { X = box.Min.X, Y = box.Max.Y, Z = z },
+            };
+
+            BH.oM.Geometry.Line rightLine = new oM.Geometry.Line
+            {
+                Start = new oM.Geometry.Point { X = box.Max.X, Y = box.Min.Y, Z = z },
+                End = new oM.Geometry.Point { X = box.Max.X, Y = box.Max.Y, Z = z },
+            };
+
+            List<BH.oM.Geometry.Line> boundarySegments = surface.ExternalBoundary.ICollapseToPolyline(BH.oM.Geometry.Tolerance.Angle).ISubParts().Select(x => x as BH.oM.Geometry.Line).ToList();
```

Comment:
This will still fail for nonlinear outlines?

Reply:
Now I see call to `ICollapseToPolyline` - plase ignore

Reply:
As discussed on Teams, `ICollapseToPolyline` should handle that :smile: